MINOR: Fix bugs in KafkaStreams.close()

jklukas opened the pull request:
Initially proposed by @ijuma in https://github.com/apache/kafka/pull/1362#issuecomment-218293662

@mjsax commented:

> StreamThread.close() should be extended to call metrics.close() (the class need a private member to reference the Metrics object, too)

The `Metrics` instance is created in the `KafkaStreams` constructor and shared between all threads, so closing it within the threads doesn't seem like the right approach. This PR calls `Metrics.close()` in `KafkaStreams.close()` instead.

cc @guozhangwang

jklukas committed:
Fix bugs in KafkaStreams.close()

ijuma commented:
The change looks good although it would be good to add a unit test. While reviewing this, I noticed that there's a bug in the `close` method. We throw an exception with a misleading message if the `state != RUNNING`. We should probably not throw an exception if `state == STOPPED`.

jklukas commented:
> it would be good to add a unit test

Will do.

> We should probably not throw an exception if state == STOPPED.

Would you like me to add that change in this PR? If so, would it be appropriate to log a WARN-level message "Method KafkaStreams.close() called on an instance that has not yet been started, so there are no threads to close"?

Although maybe the close method _should_ close the `Metrics` instance even if it's not been started. In which case, the `close` method is doing something and perhaps a warning wouldn't be appropriate.

ijuma commented:
I don't mind if it's done in this PR. We would just have to update the PR title to be a bit more general (something about fixing `KafkaStreams.close()`).

The `close` methods for the various clients are idempotent, so I think `Streams` should follow that approach as well. If `state == STOPPED`, it should mean that a previous `close` call was successful so it should just return with no warning or error.

Let's see if @guozhangwang agrees before doing the work though.

jklukas commented:
> it would be good to add a unit test

Taking a closer look at the streams tests, it seems there are currently no tests for the `KafkaStreams` class. And I'm also realizing that I'm not sure what I'd have a test do for this `close()` method.

There's some prior art in `KafkaProducerTest` that we could possibly mimic, but that would likely require creating `MockMetrics` and potentially a mock for the StreamThread array. We might not even be able to make that work because the metrics and threads members are private, so not available to substitute mocks. There may be mechanisms to inject mocks for private variables that I'm not currently aware of, though.

Is this blocked without adding a unit test for `close`? Do you any guidance about approach for adding a test?

jklukas commented:
> The close methods for the various clients are idempotent, so I think Streams should follow that approach as well

That sounds good to me.

guozhangwang commented:
re: idempotent close(): LGTM.

re: unit test: we already have a `MockMetricsReporter` for such cases, which can be passed in configs as `METRIC_REPORTER_CLASSES_CONFIG`, @jklukas you can take a look and see if that works.

guozhangwang commented:
There is one checkstyle failure: `:streams:processTestResources UP-TO-DATE
:streams:testClasses
:streams:checkstyleTest[ant:checkstyle] /home/jenkins/jenkins-slave/workspace/kafka-trunk-git-pr-jdk7/streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:33:1: Disallowed import - org.hamcrest.CoreMatchers.equalTo.
 FAILED`

ijuma commented:
Hmm, that checkstyle failure is from the bootstrap servers PR that I merged earlier today, my bad.

miguno commented:
> :streams:processTestResources UP-TO-DATE :streams:testClasses :streams:checkstyleTest[ant:checkstyle] /home/jenkins/jenkins-slave/workspace/kafka-trunk-git-pr-jdk7/streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:33:1: Disallowed import - org.hamcrest.CoreMatchers.equalTo. FAILED

Doh.  What's weird is that we use `equalTo()` elsewhere in the Streams (integration) tests, so I am confused why this is a problem here.

ijuma commented:
@guozhangwang can you please review https://github.com/apache/kafka/pull/1380? @miguno, it's explained in the other PR.

jklukas commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -198,9 +199,10 @@ public synchronized void close() {
-            throw new IllegalStateException("This process has not started yet.");
         }
+
+        metrics.close();
```
We no longer throw an exception here, and we've moved the `metrics.close` call outside the conditional so that a call to `KafkaStreams.close` will close the Metrics instance whether `start` has been called or not.

ijuma replied:
Is it safe to call this multiple times? If not, we would not want to do it if `state == STOPPED`.

jklukas commented:
Rebased on `trunk`, made `close` idempotent, added unit test based on `MockMetricsReporter`, and squashed to a single commit.

Thanks for the suggestions, and please take another look.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -0,0 +1,52 @@
+        Properties props = new Properties();
+        props.setProperty(StreamsConfig.APPLICATION_ID_CONFIG, "testStartAndClose");
+        props.setProperty(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9999");
+        props.setProperty(ProducerConfig.METRIC_REPORTER_CLASSES_CONFIG, MockMetricsReporter.class.getName());
```
There is a `StreamsConfig.METRIC_REPORTER_CLASSES_CONFIG`, we can just use that.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java:
```diff
@@ -0,0 +1,52 @@
+        props.setProperty(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9999");
+        props.setProperty(ProducerConfig.METRIC_REPORTER_CLASSES_CONFIG, MockMetricsReporter.class.getName());
+
+        final int oldInitCount = MockMetricsReporter.INIT_COUNT.get();
```
I wonder what will happen under parallel unit test environment like Jenkins, for `KafkaProducerTest`, `KafkaConsumerTest` and `KafkaStreamsTest`, won't the global counts be messed since AFAIK they are executed under a single JVM? cc @ijuma

ijuma replied:
@guozhangwang Tests are executed sequentially in each JVM. Parallel execution is achieved by forking multiple JVMs.

guozhangwang replied:
Thanks @ijuma , obviously what I learned about gradle unit tests are totally wrong.

ijuma replied:
The approach you described may also be possible (I haven't checked), but it's not what we're doing. We have enough JVM static state in some mocks that it would break lots of things if we did (as you considered in the original comment).

jklukas commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -169,8 +170,10 @@ public synchronized void start() {
+        } else if (state == RUNNING) {
             throw new IllegalStateException("This process was already started.");
+        } else {
+            throw new IllegalStateException("Cannot restart after closing.");
```
The existing error message here is only appropriate if `start()` had previously been called, but it's also possible for this to fail is `close()` is called before `start()`. There are also now tests for these two cases.

jklukas commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -194,13 +197,14 @@ public synchronized void close() {
+        }
 
+        if (state != STOPPED) {
+            metrics.close();
```
Moved `metrics.close()` to a new conditional. This should only be able to run once. Once state goes to `STOPPED`, `close` is a no-op and `start` should throw an error.

jklukas replied:
Also, added a test to ensure this section doesn't run on subsequent `close()` calls.

jklukas commented:
Addressed outstanding comments, added some new tests, and squashed. Added some inline comments about the changes to `KafkaStreams`.

guozhangwang commented:
LGTM! Merged to trunk and 0.10.0.

ijuma commented:
Thanks for the PR @jklukas, LGTM too.
